share/steel/examples/pulse/dice/dpe/DPE2.fst

```diff
@@ -0,0 +1,280 @@
+              W.release #(ht_perm cht_sig cht) cht_lk;
+
+              new_ctxt_hndl
+          }}
+          DICE_ERROR -> {
+            // ERROR -- DICE engine failed
+            free_array cdi #(coerce dice_digest_len s);
+            rewrite (engine_record_perm r r0) as (record_perm record repr);
+            fold (ht_perm sht_sig sht);
+            fold (ht_perm cht_sig cht);
+            W.release #(ht_perm sht_sig sht) sht_lk;
+            W.release #(ht_perm cht_sig cht) cht_lk;
+            0
+          }}}
+        _ -> {
+          // ERROR - repr should have type (Egnine_repr r0)
+          zeroize_array uds_len c.uds #uds_bytes;
+          disable_uds ();
+          free_array c.uds;
+
+          fold (ht_perm sht_sig sht);
+          fold (ht_perm cht_sig cht);
+          W.release #(ht_perm sht_sig sht) sht_lk;
+          W.release #(ht_perm cht_sig cht) cht_lk;
+          0
+        }}
+      }
+      _ -> {
+        // ERROR - record should have type (Engine_record r)
+        zeroize_array uds_len c.uds;
+        disable_uds ();
+        free_array c.uds;
+
+        fold (ht_perm sht_sig sht);
+        fold (ht_perm cht_sig cht);
+        W.release #(ht_perm sht_sig sht) sht_lk;
+        W.release #(ht_perm cht_sig cht) cht_lk;
+        0
+      }}
+    }
+    L0_context c -> {
+      // NOTE: we won't eventually release l0_context_perm because we won't 
+      // own it anymore -- we will free the cdi array
+      rewrite (context_perm ctxt) as (l0_context_perm c);
+      unfold (l0_context_perm c);
+
+      with s. assert (A.pts_to c.cdi full_perm s);
+
+      match record {
+      L0_record r -> {
+        match repr {
+        L0_repr r0 -> {
+          let idcsr_ing = r.deviceIDCSR_ingredients;
+          let akcrt_ing = r.aliasKeyCRT_ingredients;
+
+          let deviceIDCRI_len = len_of_deviceIDCRI  idcsr_ing.version idcsr_ing.s_common 
+                                                    idcsr_ing.s_org idcsr_ing.s_country;
+          let aliasKeyTBS_len = len_of_aliasKeyTBS  akcrt_ing.serialNumber akcrt_ing.i_common 
+                                                    akcrt_ing.i_org akcrt_ing.i_country 
+                                                    akcrt_ing.s_common akcrt_ing.s_org 
+                                                    akcrt_ing.s_country akcrt_ing.l0_version;
+          let deviceIDCSR_len = length_of_deviceIDCSR deviceIDCRI_len;
+          let aliasKeyCRT_len = length_of_aliasKeyCRT aliasKeyTBS_len;
+
+          let deviceID_pub = new_array 0uy v32us;
+          let deviceID_priv = new_array 0uy v32us;
+          let aliasKey_pub = new_array 0uy v32us;
+          let aliasKey_priv = new_array 0uy v32us;
+          let deviceIDCSR = new_array 0uy deviceIDCSR_len;
+          let aliasKeyCRT = new_array 0uy aliasKeyCRT_len;
+
+          rewrite (record_perm record repr) as (l0_record_perm r r0); 
+
+          assume_ (pure(valid_hkdf_ikm_len (digest_len dice_hash_alg)));
```

Comment:
Document this assumption? Plan to remove it?

Reply:
can likely move this into the precondition